back-end/test: add test suite for user create and delete functions

jamesytlee opened the pull request:
Adds the necessary tests for every exported function in
createFiles/createUser.js. The createUser function requires both
Mongo and Mongoose. createUser.js has also been modified for style
consistency. Also added tests for exported function in
deleteFiles/deleteUser.js. This one also requires both Mongo and
Mongoose for proper testing.

jamesytlee committed:
back-end/test: add test suite for user create and delete functions

Adds the necessary tests for every exported function in
createFiles/createUser.js. The createUser function requires both
Mongo and Mongoose. createUser.js has also been modified for style
consistency. Also added tests for exported function in
deleteFiles/deleteUser.js. This one also requires both Mongo and
Mongoose for proper testing.

forschoolandnothingelse commented on src/back-end/createFiles/createUser.js:
```diff
@@ -74,6 +78,8 @@ const createUser = (email, pwdHash, key, callback) => {
 
+/* For exporting */
 module.exports = {
+	makeid: makeid,
```
Don't export private helper functions to test them. You can test this by creating multiple users and checking for uniqueness of IDs.

forschoolandnothingelse commented on src/back-end/test/create.test.js:
```diff
@@ -0,0 +1,96 @@
+     * @param {Object} userData Object containing email and password
+     * @callback (response) Sends either true or false based on whether the user exists
+     */
+    const checkUser = async (userData, callback) => {
```
Rather than using a callback use a promise. There is no reason to have an async function with a callback.

forschoolandnothingelse reviewed (changes requested):
You should write more tests here.

Right now you don't test any error conditions. Some things you can test are
* passing invalid values as an email
* passing invalid values as a password
* passing invalid values in both

By invalid values, I mean null, undefined, etc.

forschoolandnothingelse commented on src/back-end/test/createUser.test.js:
```diff
@@ -0,0 +1,150 @@
+     * @param {Object} userData Object containing email and password
+     * @return true if user exists, false otherwise
+     */
+    const checkUser = async (userData) => {
```
You shouldn't be catching errors in this function.
You also don't return anything in the case that it is true.

forschoolandnothingelse commented on src/back-end/test/createUser.test.js:
```diff
@@ -0,0 +1,150 @@
+            email: EMAIL,
+            pwd: PWDHASH
+        };
+        await createUser(EMAIL, PWDHASH, KEY, () => {
```
This function doesn't take a callback anymore. Your test never gets executed.
You should look up wait `await` does.
Basically the test should look something like:
```js
let user = await createUser(EMAIL, PWDHAS, KEY);
let success = checkUser(user);
expect(success).toBe(true);
```

forschoolandnothingelse commented on src/back-end/test/createUser.test.js:
```diff
@@ -0,0 +1,150 @@
+        const PWDHASH2 = "456DEF";
+        const KEY1 = "ABC123";
+        const KEY2 = "ABC123";
+        await createUser(EMAIL1, PWDHASH1, KEY1, () => {
```
Same problem, this doesn't test anything because this function doesn't take callbacks.

forschoolandnothingelse commented on src/back-end/test/createUser.test.js:
```diff
@@ -0,0 +1,150 @@
+            email: EMAIL,
+            pwd: PWDHASH2
+        };
+        await createUser(EMAIL, PWDHASH1, KEY, () => {
```
No callbacks

forschoolandnothingelse commented on src/back-end/test/createUser.test.js:
```diff
@@ -0,0 +1,150 @@
+            email: EMAIL,
+            pwd: PWDHASH
+        };
+        await createUser(EMAIL, PWDHASH, KEY, () => {
```
No callbacks

forschoolandnothingelse commented on src/back-end/test/createUser.test.js:
```diff
@@ -0,0 +1,150 @@
+            email: EMAIL,
+            pwd: PWDHASH
+        };
+        await createUser(EMAIL, PWDHASH, KEY, () => {
```
No callbacks

forschoolandnothingelse commented on src/back-end/test/createUser.test.js:
```diff
@@ -0,0 +1,150 @@
+            email: EMAIL,
+            pwd: PWDHASH
+        };
+        await createUser(EMAIL, PWDHASH, KEY, () => {
```
No callbacks

forschoolandnothingelse commented on src/back-end/test/deleteUser.test.js:
```diff
@@ -0,0 +1,80 @@
+     * @param {Object} userData Object containing email and password
+     * @return true if user exists, false otherwise
+     */
+     const checkUser = async (userData) => {
```
Don't catch errors

forschoolandnothingelse commented on src/back-end/test/deleteUser.test.js:
```diff
@@ -0,0 +1,80 @@
+            email: EMAIL,
+            pwd: PWDHASH
+        };
+        await createUser(EMAIL, PWDHASH, KEY, () => {
```
No callbacks

forschoolandnothingelse commented on src/back-end/test/deleteUser.test.js:
```diff
@@ -0,0 +1,80 @@
+            let success = checkUser(user);
+            expect(success).toBe(true);
+        });
+        await deleteUser(EMAIL, () => {
```
No callbacks

forschoolandnothingelse commented on src/back-end/test/deleteUser.test.js:
```diff
@@ -0,0 +1,80 @@
+
+    test("Non-existing user cannot be deleted", async () => {
+        const EMAIL = "foo@gmail.com";
+        await deleteUser(EMAIL).
```
If you're going to use `then` and `catch` you need to have just pass in `done` and call it at the end of `then` and `catch`. No need for `await` or for this function to be marked `async`.

jamesytlee committed:
back-end/createUser: check for invalid email

Small change that checks for invalid emails,
which includes empty strings, null values,
and emails missing '@' character.

jamesytlee committed:
back-end/createUser: changes for password argument

Modified createUser to check for invalid password
lengths. Also changed index.js to fit the changes of
createUser. Now the createUser function does not take
a hashed password as its argument, and rather the regular
password.

jamesytlee committed:
back-end/test/security: fixes non-deterministic tests

Small change to allow deterministic tests. Previously, at
random times, the test would not pass. This is no longer the
case.

forschoolandnothingelse commented:
I finished a pair programming session with James and we fixed all the issues with this PR. All tests pass and design changes made are reasonable. Approved.
